File: 04.GrafosIntro/bfs.py
import sys
from queue import Queue
from graph import Graph

class BFS:

    def __init__(self, g, s):
        self.g = g
        self.start = s

        self.marked = {}
        self.edgeTo = {}
        self.distTo = {}

        self.bfs(s)

    def bfs(self, v):
        fila = Queue()
        fila.put(v)
        while fila.qsize() > 0:
            for w in g.getAdj(v):
                if w not in self.marked:
                    pass

    def hasPathTo(self, v):
        return v in self.marked
    
    def pathTo(self, v):
        path = []
        if not self.hasPathTo(v):
            return path
        while v != self.start:
            path.insert(0,v)
            v = self.edgeTo[v]
        path.insert(0,self.start)
        return path

if __name__ == "__main__":

    g = Graph("exemplos/tinyG.txt")
    bfs = BFS(g, "0")

    for v in g.getVerts():
        print(f"{v}: ", end="")
        if bfs.hasPathTo(v):
            for w in bfs.pathTo(v):
                print(f"{w} ", end="")
            print()
        print()
    print()

File: 04.GrafosIntro/BFS.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class BFS {
    private Set<String> marked; // vértices já visitados
    private Map<String, String> edgeTo; // antecessores de cada vértice na árvore de caminhos
    private Map<String, Integer> distTo; // distância em arestas do start até cada vértice
    private String start; // vértice inicial

    public BFS(Graph g, String start) {
        this.start = start;
        marked = new HashSet<>();
        edgeTo = new HashMap<>();
        distTo = new HashMap<>();
        bfs(g, start);
    }

    private void bfs(Graph g, String s) {
        LinkedList<String> fila = new LinkedList<>();
        // System.out.println(">>> Entrando em " + s);
        marked.add(s); // "marca" como visitado
        // Para cada vizinho w de s...
        for (String w : g.getAdj(s)) {
            // Se não estiver marcado...
            if (!marked.contains(w)) {
                // System.out.println(">>> Vou visitar " + w + " a partir de " + s);
                // edgeTo.put(w, s); // indica que para chegar em w, viemos de s
                // dfs(g, w);
            } else {
                // System.out.println("!!! Não visito " + w + " pois já está marcado");
            }
        }
        // System.out.println(">>> Saindo de " + s);
    }

    // Retorna true se há caminho a partir de start até v
    public boolean hasPathTo(String v) {
        if (marked.contains(v))
            return true;
        return false;
    }

    // Retorna o caminho de start até v
    public List<String> pathTo(String v) {
        LinkedList<String> caminho = new LinkedList<>();
        if (!hasPathTo(v))
            return caminho; // retorna lista vazia se não houver caminho
        while (!v.equals(start)) {
            caminho.addFirst(v);
            v = edgeTo.get(v);
        }
        caminho.addFirst(start); // ou v
        return caminho;
    }

    public static void main(String[] args) {
        Graph g = new Graph("exemplos/tinyG.txt");
        BFS bfs = new BFS(g, "0");

        // Mostra todos os caminhos encontrados
        for (String v : g.getVerts()) {
            System.out.print(v + ": ");
            if (bfs.hasPathTo(v)) {
                for (String w : bfs.pathTo(v)) {
                    System.out.print(w + " ");
                }
                System.out.println();
            } else
                System.out.println("não tem caminho...");
        }
    }
}
